### DISCUSSION TITLE
Object is mistakenly treated as array

### DISCUSSION THREAD

**Discussion by LiadAvramov** [nlohmann/json#4441] [Q&A]
I have this simple json object:

```
json obj, error;
error["error"] = "error_str";
obj["key"] = error;
```

That is to be deserialized as follow:
{"key": {"error":"error_str"}}

When I try to push this json into a std::vector:

std::vector<json> artifacts;
`artifacts.push_back(obj["key"]);`

I get the following exception:

terminate called after throwing an instance of 'nlohmann::json_abi_v3_11_3::detail::type_error'
  what():  [json.exception.type_error.305] cannot use operator[] with a string argument with array

I did some tests and apparently the inner json (ie. {"error":"error_str"}) is treated as an array.
My question is how can it possibly be treated that way? What am I doing wrong ?

**Comment by nlohmann** [comment]
The posted snippet doesn't actually reproduce the described exception: `json obj, error; error["error"] = "error_str"; obj["key"] = error;` correctly produces `{"key":{"error":"error_str"}}`, and pushing `obj["key"]` into a `std::vector<json>` just copies that object value with no issue. The type_error.305 ('cannot use operator[] with a string argument with array') means some json value really is holding an array at the point of that string-key access -- most likely `artifacts`'s element type isn't plain `nlohmann::json` but a custom struct with its own from_json/adl_serializer expecting an array, or one of `obj`/`error`/`artifacts` is reassigned elsewhere in the real (not-shown) code. Sharing the actual declaration of `artifacts` and any custom to_json/from_json involved would be needed to pin down the real cause.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
